### ISSUE TITLE
Feature: Allow listen to accept multiple connections, spawning for each

### ISSUE THREAD

**Issue by Grazfather** [Gallopsled/pwntools#742]
Right now on accept() it assigns to self.sock from the accepter thread, so this would require an overhaul there to support more than one.

**Comment by zachriggle** [comment]
Correct, the `listen()` tube is currently one-shot by design.  I've always wondered about this design choice.  It seems better to use:

``` py
l = listen()
a = l.wait_for_connection()
b = l.wait_for_connection()
c = l.wait_for_connection()
```

Such that `l` continues listening for additional connections on each call to `wait_for_connection`.

Since this would break backward compatibility, I'd suggest adding a new tube named `server`, or adding an optional argument (e.g. `server=True`, default value of `False`) in order to keep waiting.

Additionally, it might be useful to provide a callback for new connections, which are invoked in a new thread.  This will require one thread for the `listen()` or `server()` object, and an additional thread per-connection (though perhaps we can force it to handle requests serially?).

What do you think?

**Comment by Grazfather** [comment]
That's actually exactly how I was using it, and it worked fine, but kept returning the same connection. (And it took me too long to notice this).

I won't weigh in on `server` vs `listen(server=True)`. A callback would be nice, and optionally maybe a collection of them. In my particular case (for CSAW2016  mom's spaghetti, fwiw), I need a few connections up quickly and need each to have different behaviour.

I can't even easily just spawn x listeners obviously since the first is bound to the port.

**Comment by zachriggle** [comment]
You can do it trivially under the current circumstance as long as you can:
1. Use different ports
2. Accept connections serially

For example, this generates 100 one-shot listeners on a bunch of ports.

``` py
listeners = map(lambda: listen(), 100)
for l in listeners:
    print l.lport
```

Separately, you can reuse the same port after the connection has been accepted.

``` py
port = 4141
conns = []
for i in range(10):
  conns.append(listen(port).wait_for_connection())
```

**Comment by Grazfather** [comment]
Yeah actually I tried almost exactly that second example. It might have worked remotely, but when running the callback service locally I think the two `connect`s come too quickly.

``` bash
[pid 18741] socket(PF_INET, SOCK_STREAM, IPPROTO_IP) = 5
[pid 18741] connect(5, {sa_family=AF_INET, sin_port=htons(55556), sin_addr=inet_addr("127.0.0.1")}, 16) = 0
[pid 18741] recv(5, Process 18740 attached
 <unfinished ...>
...
[pid 18740] socket(PF_INET, SOCK_STREAM, IPPROTO_IP) = 6
[pid 18740] connect(6, {sa_family=AF_INET, sin_port=htons(55556), sin_addr=inet_addr("127.0.0.1")}, 16) = 0
[pid 18740] recv(6, 0xf6200468, 8, 0)   = -1 ECONNRESET (Connection reset by peer)
```

I think what is happening here is that the first listener isn't unbinding in time (it has to wait for the gc?), so the second incoming connection is connecting and then disconnecting immediately.

I should be able to handle this with raw sockets, though, just thought it'd be nice for pwntools.

**Comment by devneal** [comment]
I submitted a pull request at #1067 that works for my purposes, but any comments on robustness for the main branch would be appreciated.

**Comment by Arusekk** [comment]
Closed by #1067 #1707

**Comment by Grazfather** [comment]
Sweet I can finally solve the challenge!
